crates/net/network/src/transactions/fetcher.rs

```diff
@@ -816,76 +827,118 @@ impl TransactionFetcher {
 `%peer_id`: {},
 `@self`: {:?}",
                 peer_id, self
             );
 
-            self.decrement_inflight_request_count_for(&peer_id);
-
-            return match result {
-                Ok(Ok(transactions)) => {
-                    // clear received hashes
-                    let mut fetched = Vec::with_capacity(transactions.len());
-                    requested_hashes.retain(|requested_hash| {
-                        if transactions.hashes().any(|hash| hash == requested_hash) {
-                            // hash is now known, stop tracking
-                            fetched.push(*requested_hash);
-                            return false
-                        }
-                        true
-                    });
-                    fetched.shrink_to_fit();
+        self.decrement_inflight_request_count_for(&peer_id);
 
-                    self.remove_hashes_from_transaction_fetcher(fetched);
+        match result {
+            Ok(Ok(transactions)) => {
+                let payload = UnverifiedPooledTransactions::new(transactions);
 
-                    // buffer left over hashes
-                    self.buffer_hashes_for_retry(requested_hashes, &peer_id);
+                let unverified_len = payload.len();
+                let (verification_outcome, verified_payload) =
+                    payload.verify(&requested_hashes, &peer_id);
 
-                    Poll::Ready(Some(FetchEvent::TransactionsFetched {
-                        peer_id,
-                        transactions: transactions.0,
-                    }))
-                }
-                Ok(Err(req_err)) => {
-                    self.buffer_hashes_for_retry(requested_hashes, &peer_id);
-                    Poll::Ready(Some(FetchEvent::FetchError { peer_id, error: req_err }))
+                if let VerificationOutcome::ReportPeer = verification_outcome {
+                    // todo: report peer for sending hashes that weren't requested
+                    trace!(target: "net::tx",
+                        peer_id=format!("{peer_id:#}"),
+                        unverified_len=unverified_len,
+                        verified_payload_len=verified_payload.len(),
+                        "received `PooledTransactions` response from peer with entries that didn't verify against request, filtered out transactions"
+                    );
                 }
-                Err(_) => {
-                    self.buffer_hashes_for_retry(requested_hashes, &peer_id);
-                    // request channel closed/dropped
-                    Poll::Ready(Some(FetchEvent::FetchError {
-                        peer_id,
-                        error: RequestError::ChannelClosed,
-                    }))
+
+                let unvalidated_payload_len = verified_payload.len();
+
+                // todo: report peer for sending invalid response
```

Comment:
is this for a follow up?

Reply:
yeah exactly can add which issues it depends on